makeflow/src/makeflow_hook.h

```diff
@@ -0,0 +1,322 @@
+	 */
+	int (*create)        (struct jx *);
+
+	/* Destroy/Clean up hooks.
+	 *
+	 * Call when Makeflow is about to exit. This is to clean up any memory
+	 * or structs left during execution. Most will be removed when exiting.
+	 * This should only be used to clean up internal strucutures.
+	 *
+	 * @return 1 if successfully destroyed, 0 if not.
+	 */
+	int (*destroy)       ();
+
+	/* Hook prior to dag creation.
+	 * 
+	 * This is set after hook create, but prior to DAG creation.
+	 * This is for adding features to the Makeflow environment, but not to
+	 * the dag itself as it does not exist.
+	 *
+	 * @return 1 if dag init step successful, 0 if not.
+	 */
+	int (*dag_init)      ();
+
+	/* Hook prior to dag start, but after parsing.
+	 * 
+	 * This is used to augment the DAG or utilize information from the
+	 * DAG to make decisions. An example of this is the storage
+	 * allocation module that looks at file information that is parsed.
+	 *
+	 * @param dag The DAG about to be started.
+	 * @return 1 if dag start step successful, 0 if not.
+	 */
+	int (*dag_start)     (struct dag *);
+
+	/* Hook for a successfully completed DAG.
+	 * 
+	 * @param dag The DAG that was complete.
+	 * @return 1 if dag end step successful, 0 if not.
+	 */
+	int (*dag_end)       (struct dag *);
+
+	/* Hook for a failed DAG.
+	 * 
+	 * This does not change that the DAG has failed, but gives the
+	 * hook access to internal stats for failure analysis.
+	 *
+	 * @param dag The DAG that was failed.
+	 * @return 1 if dag fail step successful, 0 if not.
+	 */
+	int (*dag_fail)      (struct dag *);
+
+	/* Hook for an aborted DAG.
+	 * 
+	 * This does not change that the DAG has aborted, but gives the
+	 * hook access to internal stats for abort analysis.
+	 *
+	 * @param dag The DAG that was aborted.
+	 * @return 1 if dag abort step successful, 0 if not.
+	 */
+	int (*dag_abort)     (struct dag *);
+
+	/* ADD WRAPPERS IN EITHER CREATE CHECK OR SUBMIT */
+
+	/* Hook when a node is created.
+	 * 
+	 * This hook occurs during parse when a node is created. Is the first
+	 * opportunity to see the command, files, env, and resources.
+	 *
+	 * @param dag_node The dag_node that was just created.
+	 * @param batch_job_feature A strucuture that describes supported
+	 *             features of used batch_job system.
+	 * @return 1 if successful, 0 if not.
+	 */
+	int (*node_create)   (struct dag_node *, struct hash_table *);
+
+	/* Hook when a node is checked for submission.
+	 * 
+	 * This hook occurs when nodes are checked for execution. This
+	 * allows hooks to veto submission based on internal qualifiers.
+	 * Example would be the storage allocation, or job limits.
+	 *
+	 * @param dag_node The dag_node that is being checked.
+	 * @param batch_job_feature A strucuture that describes supported
+	 *             features of used batch_job system.
+	 * @return 1 if successful, 0 if not.
+	 */
+	int (*node_check)    (struct dag_node *, struct hash_table *);
```

Comment:
Since the doc comments refer to parameters, it might be good to add dummy names in the header definitions, e.g.

    int (*node_check) (struct dag_node *d, struct hash_table *f);

so the names in the comments line up. Here, `batch_job_feature` isn't obviously connected with a `struct 
hash_table *`.